### ISSUE TITLE
BadMethodCallException: The option "base_path" does not have a callable "setBasePath"

### ISSUE THREAD

**Issue by aiham** [shlinkio/shlink#880]
#### How Shlink is set-up

* Shlink Version: 2.0.5
* PHP Version: 7.4.1
* How do you serve Shlink: Docker image (shlinkio/shlink:2.0.5)
* Database engine used: SQLite
* nginx version: nginx/1.14.0 (Ubuntu 18.04)

#### Summary

An unexpected error is thrown when visiting the root path and random unknown paths. Short URLs do successfully redirect to long URLs so no problem there, it's just for the root path (https://abc.co) and unknown paths (https://abc.co/some-random-path).

#### Current behavior

- Returns a 500 error with this stack trace in the docker logs:

```
shlink_1  | [2020-11-01T18:25:38.038998+00:00] Shlink.ERROR - Laminas\Stdlib\Exception\BadMethodCallException: The option "base_path" does not have a callable "setBasePath" ("setbasepath") setter method which must be defined in /etc/shlink/vendor/laminas/laminas-stdlib/src/AbstractOptions.php:111
shlink_1  | Stack trace:
shlink_1  | #0 /etc/shlink/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(63): Laminas\Stdlib\AbstractOptions->__set('base_path', 'https://www.foo...')
shlink_1  | #1 /etc/shlink/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(33): Laminas\Stdlib\AbstractOptions->setFromArray(Array)
shlink_1  | #2 /etc/shlink/vendor/laminas/laminas-servicemanager/src/AbstractFactory/ConfigAbstractFactory.php(71): Laminas\Stdlib\AbstractOptions->__construct(Array)
shlink_1  | #3 /etc/shlink/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(765): Laminas\ServiceManager\AbstractFactory\ConfigAbstractFactory->__invoke(Object(Laminas\ServiceManager\ServiceManager), 'Shlinkio\\Shlink...', NULL)
shlink_1  | #4 /etc/shlink/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(201): Laminas\ServiceManager\ServiceManager->doCreate('Shlinkio\\Shlink...')
shlink_1  | #5 [internal function]: Laminas\ServiceManager\ServiceManager->get('Shlinkio\\Shlink...')
shlink_1  | #6 /etc/shlink/vendor/laminas/laminas-servicemanager/src/AbstractFactory/ConfigAbstractFactory.php(69): array_map(Array, Array)
shlink_1  | #7 /etc/shlink/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(765): Laminas\ServiceManager\AbstractFactory\ConfigAbstractFactory->__invoke(Object(Laminas\ServiceManager\ServiceManager), 'Shlinkio\\Shlink...', NULL)
shlink_1  | #8 /etc/shlink/vendor/laminas/laminas-servicemanager/src/ServiceManager.php(201): Laminas\ServiceManager\ServiceManager->doCreate('Shlinkio\\Shlink...')
shlink_1  | #9 /etc/shlink/vendor/mezzio/mezzio/src/MiddlewareContainer.php(64): Laminas\ServiceManager\ServiceManager->get('Shlinkio\\Shlink...')
shlink_1  | #10 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(46): Mezzio\MiddlewareContainer->get('Shlinkio\\Shlink...')
shlink_1  | #11 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #12 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(84): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #13 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #14 /etc/shlink/vendor/laminas/laminas-stratigility/src/Middleware/PathMiddlewareDecorator.php(48): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #15 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Laminas\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #16 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(54): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #17 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(54): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #18 /etc/shlink/module/Core/src/Action/RedirectAction.php(23): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #19 /etc/shlink/module/Core/src/Action/AbstractTrackingAction.php(63): Shlinkio\Shlink\Core\Action\RedirectAction->createErrorResp(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #20 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(47): Shlinkio\Shlink\Core\Action\AbstractTrackingAction->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #21 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #22 /etc/shlink/vendor/akrabat/ip-address-middleware/src/IpAddress.php(91): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #23 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(47): RKA\Middleware\IpAddress->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #24 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #25 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(84): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #26 /etc/shlink/vendor/mezzio/mezzio-router/src/Route.php(101): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #27 /etc/shlink/vendor/mezzio/mezzio-router/src/RouteResult.php(111): Mezzio\Router\Route->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #28 /etc/shlink/vendor/mezzio/mezzio-router/src/Middleware/DispatchMiddleware.php(36): Mezzio\Router\RouteResult->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))shlink_1  | #29 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(47): Mezzio\Router\Middleware\DispatchMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #30 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Mezzio\Middleware\LazyLoadingMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #31 /etc/shlink/vendor/laminas/laminas-stratigility/src/MiddlewarePipe.php(84): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #32 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Laminas\Stratigility\MiddlewarePipe->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #33 /etc/shlink/vendor/laminas/laminas-stratigility/src/Middleware/PathMiddlewareDecorator.php(48): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #34 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(61): Laminas\Stratigility\Middleware\PathMiddlewareDecorator->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
shlink_1  | #35 /etc/shlink/vendor/laminas/laminas-stratigility/src/Next.php(54): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #36 /etc/shlink/vendor/mezzio/mezzio-router/src/Middleware/RouteMiddleware.php(55): Laminas\Stratigility\Next->handle(Object(Laminas\Diactoros\ServerRequest))
shlink_1  | #37 /etc/shlink/vendor/mezzio/mezzio/src/Middleware/LazyLoadingMiddleware.php(47): Mezzio\Router\Middleware\RouteMiddleware->process(Object(Laminas\Diactoros\ServerRequest), Object(Laminas\Stratigility\Next))
``` 

#### Expected behavior

- Should redirect to the configured `base_url_redirect_to` or `regular_404_redirect_to`

#### How to reproduce

1. Run docker image shlinkio/shlink:2.0.5
2. Mount this JSON config in a volume on the container:

```
{
    "disable_track_param": "__my_disable_param",
    "delete_short_url_threshold": 0,
    "short_domain_schema": "https",
    "short_domain_host": "abc.co",
    "validate_url": false,
    "invalid_short_url_redirect_to": "https://www.foobar.com",
    "regular_404_redirect_to": "https://www.foobar.com",
    "base_url_redirect_to": "https://www.foobar.com",
    "default_short_codes_length": 4,
    "anonymize_remote_addr": false,
    "geolite_license_key": "fake_key",
    "port": 8888
}
```

I tried without any `base_path`, with `"base_path": ""` and with `"base_path": "/"`, all had the same result.

3. Expose an nginx service to the internet and proxy pass to the docker container:

```
server {
    listen 443 ssl http2;
    listen [::]:443 ssl http2;

    server_name abc.co;

    ssl_certificate /etc/nginx/certs/certificate.crt;
    ssl_certificate_key /etc/nginx/certs/certificate.key;

    location / {
        proxy_pass http://127.0.0.1:8080;
        proxy_pass_header Server;
        proxy_read_timeout 90s;

        proxy_redirect    off;
        proxy_set_header  Host             $http_host;
        proxy_set_header  X-Real-IP        $remote_addr;
        proxy_set_header  X-Forwarded-For  $proxy_add_x_forwarded_for;
        proxy_set_header  X-Forwarded-Proto $scheme;
    }

    server_tokens off;
}
```

I think this is probably the important line:

```
shlink_1  | #0 /etc/shlink/vendor/laminas/laminas-stdlib/src/AbstractOptions.php(63): Laminas\Stdlib\AbstractOptions->__set('base_path', 'https://www.foo...')
```

I haven't checked the source code but it doesn't seem right that it's setting the base path to the full URL (probably comes from from one of the three `_redirect_to` config values). Could this be because of the nginx proxy?

**Comment by aiham** [comment]
I just found https://github.com/shlinkio/shlink/issues/665
I'll try the workaround with the ENV var. Not sure if the submitted fix is deployed as a container yet, I'm using stable (2.0.5), i don't see a 2.1.0 build on docker hub.

**Comment by aiham** [comment]
The workaround worked. Would be good to have the latest version deployed to docker hub. Thanks for building shlink.
Cheers

**Comment by acelaya** [comment]
Hey @aiham 

This was a bug that was fixed on v2.1.0

If you update to that version or any newer one, you should be fine.

Every version is published in docker hub automatically. If it looks like it's not there is because docker hub sometimes does not properly sort all existing tags.

* [2.1](https://hub.docker.com/layers/shlinkio/shlink/2.1.0/images/sha256-0b98145639dda497da5abfc492e353f69d78c1dae1c6361a007774dd50a84db7?context=explore)
* [2.2](https://hub.docker.com/layers/shlinkio/shlink/2.2.0/images/sha256-8da5c634f5a3fc4a2d45d4a2cc4c4e4aea3bd290f9fbe4f48b52fe0004bbb08b?context=explore)
* [2.3](https://hub.docker.com/layers/shlinkio/shlink/2.3.0/images/sha256-252c6ba4f7e46fb48af05accc40cc10f24851ae29914c13e23afa502721354be?context=explore)

The different patches for every version are also available. Basically any tag you see in this repository has a corresponding docker tag.

**Comment by aiham** [comment]
Thanks for the response. Yeah looks like that's exactly what happened, I went through the first few pages looking for 2.1.0 (I saw the PR was tagged for that release) but 2.0.5 was the highest I found. Thanks for pointing that out.
